File: backend-go/projects/gateway/internal/gatewaypreauth/modelsresponseprotocol.go
package gatewaypreauth

import (
	"regexp"
	"strings"
)

// Port of request/models-response-protocol.ts: which protocol serves the
// GET /models request when the authenticated models fast path applies.

// ResponseProtocolCode mirrors the response protocol union values used here.
type ResponseProtocolCode string

const (
	ResponseProtocolOpenAI     ResponseProtocolCode = "openai_v1"
	ResponseProtocolAnthropicV ResponseProtocolCode = "anthropic_v1"
	ResponseProtocolGeminiV    ResponseProtocolCode = "gemini_v1beta"
)

// GatewayClientProfileHeader mirrors gatewayClientProfileHeader.
const GatewayClientProfileHeader = "x-juhe-client-profile"

// ResolveGatewayModelsResponseProtocol mirrors resolveGatewayModelsResponseProtocol.
func ResolveGatewayModelsResponseProtocol(req *GatewayRequest) (ResponseProtocolCode, bool) {
	if IsGeminiModelsRequest(req) && isExplicitGeminiModelsClient(req) {
		return ResponseProtocolGeminiV, true
	}
	if IsAnthropicModelsRequest(req) && isExplicitAnthropicModelsClient(req) {
		return ResponseProtocolAnthropicV, true
	}
	if IsOpenAIModelsRequest(req) || IsAnthropicModelsRequest(req) || IsGeminiModelsRequest(req) {
		return ResponseProtocolOpenAI, true
	}
	return "", false
}

// isExplicitGeminiModelsClient mirrors the Node helper field by field.
func isExplicitGeminiModelsClient(req *GatewayRequest) bool {
	path, query := splitPathAndQuery(req.PathAndQuery())
	if strings.ToLower(path) == "/v1beta/models" {
		return true
	}
	profile := normalizedHeaderToken(req.Header(GatewayClientProfileHeader))
	if profile == "gemini" || profile == "generic_gemini" || profile == "gemini_cli" {
		return true
	}
	if lowerHeaderToken(req.Header("x-goog-api-key")) != "" {
		return true
	}
	if key, ok := queryParamFirstValue(query, "key"); ok && strings.TrimSpace(key) != "" {
		return true
	}
	userAgent := lowerHeaderToken(req.Header("user-agent"))
	return userAgent != "" && (geminiCLIUserAgentPattern.MatchString(userAgent) || geminiCLIProxyPattern.MatchString(userAgent))
}

var (
	geminiCLIUserAgentPattern = regexp.MustCompile(`(?i)\bgeminicli(?:[-/]|$)`)
	geminiCLIProxyPattern     = regexp.MustCompile(`(?i)\bproxy_client=geminicli\b`)
)

// isExplicitAnthropicModelsClient mirrors the Node helper.
func isExplicitAnthropicModelsClient(req *GatewayRequest) bool {
	profile := normalizedHeaderToken(req.Header(GatewayClientProfileHeader))
	if profile == "anthropic" || profile == "generic_anthropic" || profile == "claude_code" {
		return true
	}
	if normalizedHeaderToken(req.Header("anthropic-version")) != "" ||
		normalizedHeaderToken(req.Header("anthropic-beta")) != "" ||
		normalizedHeaderToken(req.Header("x-claude-code-session-id")) != "" ||
		normalizedHeaderToken(req.Header("x-claude-code-agent-id")) != "" {
		return true
	}
	return claudeCodeUserAgent(req)
}

// claudeCodeUserAgent mirrors claudeCodeUserAgent.
func claudeCodeUserAgent(req *GatewayRequest) bool {
	userAgent := lowerHeaderToken(req.Header("user-agent"))
	return userAgent != "" && (strings.HasPrefix(userAgent, "claude-cli/") || strings.Contains(userAgent, " claude-cli/"))
}

// normalizedHeaderToken mirrors normalizedHeaderToken: trim, lowercase and
// collapse runs of '-' and whitespace into '_'.
func normalizedHeaderToken(value string) string {
	lower := lowerHeaderToken(value)
	if lower == "" {
		return ""
	}
	return collapseSeparators(lower)
}

// lowerHeaderToken mirrors lowerHeaderToken: trimmed lowercase or ”.
func lowerHeaderToken(value string) string {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return ""
	}
	return strings.ToLower(trimmed)
}

// collapseSeparators mirrors replace(/[-\s]+/g, '_').
func collapseSeparators(value string) string {
	var builder strings.Builder
	previousSeparator := false
	for _, r := range value {
		if r == '-' || r == ' ' || r == '\t' || r == '\n' || r == '\r' || r == '\v' || r == '\f' {
			if !previousSeparator {
				builder.WriteByte('_')
			}
			previousSeparator = true
			continue
		}
		previousSeparator = false
		builder.WriteRune(r)
	}
	return builder.String()
}

File: migration-backup-1/node/final-archive/backend/src/modules/gateway/request/models-response-protocol.ts
import type { Request } from 'express'

import { gatewayClientProfileHeader } from '../client-profiles/strategy.js'
import { isAnthropicModelsRequest } from '../protocols/anthropic-v1/route-helpers.js'
import { isGeminiModelsRequest } from '../protocols/gemini-v1beta/route-helpers.js'
import { isOpenAIModelsRequest, splitPathAndQuery } from '../protocols/openai-v1/route-helpers.js'
import type { ResponseProtocolCode } from '../protocols/openai-v1/response-semantics.js'

export function resolveGatewayModelsResponseProtocol(req: Request): ResponseProtocolCode | undefined {
  if (isGeminiModelsRequest(req) && isExplicitGeminiModelsClient(req)) {
    return 'gemini_v1beta'
  }
  if (isAnthropicModelsRequest(req) && isExplicitAnthropicModelsClient(req)) {
    return 'anthropic_v1'
  }
  if (isOpenAIModelsRequest(req) || isAnthropicModelsRequest(req) || isGeminiModelsRequest(req)) {
    return 'openai_v1'
  }
  return undefined
}

function isExplicitGeminiModelsClient(req: Request): boolean {
  const { path, query } = splitPathAndQuery(req.originalUrl || req.path)
  if (path.toLowerCase() === '/v1beta/models') return true

  const profile = normalizedHeaderToken(req.header(gatewayClientProfileHeader))
  if (profile === 'gemini' || profile === 'generic_gemini' || profile === 'gemini_cli') {
    return true
  }
  if (lowerHeaderToken(req.header('x-goog-api-key'))) return true
  const key = new URLSearchParams(query.startsWith('?') ? query.slice(1) : query).get('key')
  if (typeof key === 'string' && key.trim()) return true

  const userAgent = lowerHeaderToken(req.header('user-agent'))
  return Boolean(userAgent && (/\bgeminicli(?:[-/]|$)/i.test(userAgent) || /proxy_client=geminicli\b/i.test(userAgent)))
}

function isExplicitAnthropicModelsClient(req: Request): boolean {
  const profile = normalizedHeaderToken(req.header(gatewayClientProfileHeader))
  if (profile === 'anthropic' || profile === 'generic_anthropic' || profile === 'claude_code') {
    return true
  }
  return Boolean(
    normalizedHeaderToken(req.header('anthropic-version'))
      || normalizedHeaderToken(req.header('anthropic-beta'))
      || normalizedHeaderToken(req.header('x-claude-code-session-id'))
      || normalizedHeaderToken(req.header('x-claude-code-agent-id'))
      || claudeCodeUserAgent(req)
  )
}

function claudeCodeUserAgent(req: Request): boolean {
  const userAgent = lowerHeaderToken(req.header('user-agent'))
  return Boolean(userAgent && (userAgent.startsWith('claude-cli/') || userAgent.includes(' claude-cli/')))
}

function normalizedHeaderToken(value: unknown): string | undefined {
  return lowerHeaderToken(value)?.replace(/[-\s]+/g, '_')
}

function lowerHeaderToken(value: unknown): string | undefined {
  return typeof value === 'string' && value.trim()
    ? value.trim().toLowerCase()
    : undefined
}
